source/common/router/router.cc

```diff
@@ -749,13 +749,21 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
                                         can_use_http3, false /*enable_half_close*/);
   if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() && ensure_connection_retry_) {
     upstream_request->addUpstreamCallbacks(*this);
-    wait_for_connect_ = true;
-    // Set to the maximum between the current limit and 2 time the cluster connection buffer limit.
-    // 2 times in order to ensure we do not reach this limit until the connection get establish,
-    // the upstream request will pause the read on the downstream connection but this is a soft
-    // limit, so it might goes above this limit.
+    waiting_for_upstream_connection_ = true;
+    // Set to the maximum between the current limit and the cluster connection buffer limit plus
+    // 1 MiB. That is to ensure we do not reach this limit until the connection get establish, the
+    // upstream request will pause the read on the downstream connection but this is a soft limit,
+    // so it might goes above this limit.
+    uint32_t cluster_connection_buffer_limit = cluster_->perConnectionBufferLimitBytes();
+    // Ensure no integer overflow.
+    if (cluster_connection_buffer_limit < std::numeric_limits<uint32_t>::max() - 1048576) {
```

Comment:
If we stick with +1m we could set to std::max (limit, limit +1m) rather than if/else and hoping the numbers are in sync.
I still don't understand why we can't use the per connection buffer limit though

Reply:
done